Flight delay Predictor - Data Wrangling

In [ ]:
#Import Libraries 

import pandas as pd
import numpy as np

In [3]:
# Load dataset
df = pd.read_csv("Data/Airline_Delay_Cause.csv")

# Preview data
df.head()

year  month carrier       carrier_name airport  \
0  2023      8      9E  Endeavor Air Inc.     ABE   
1  2023      8      9E  Endeavor Air Inc.     ABY   
2  2023      8      9E  Endeavor Air Inc.     AEX   
3  2023      8      9E  Endeavor Air Inc.     AGS   
4  2023      8      9E  Endeavor Air Inc.     ALB   

                                        airport_name  arr_flights  arr_del15  \
0  Allentown/Bethlehem/Easton, PA: Lehigh Valley ...         89.0       13.0   
1             Albany, GA: Southwest Georgia Regional         62.0       10.0   
2           Alexandria, LA: Alexandria International         62.0       10.0   
3        Augusta, GA: Augusta Regional at Bush Field         66.0       12.0   
4                   Albany, NY: Albany International         92.0       22.0   

   carrier_ct  weather_ct  ...  security_ct  late_aircraft_ct  arr_cancelled  \
0        2.25        1.60  ...          0.0              5.99            2.0   
1        1.97        0.04  ...          0.0              7.42            0.0   
2        2.73        1.18  ...          0.0              4.28            1.0   
3        3.69        2.27  ...          0.0              1.57            1.0   
4        7.76        0.00  ...          0.0             11.28            2.0   

   arr_diverted  arr_delay  carrier_delay  weather_delay  nas_delay  \
0           1.0     1375.0           71.0          761.0      118.0   
1           1.0      799.0          218.0            1.0       62.0   
2           0.0      766.0           56.0          188.0       78.0   
3           1.0     1397.0          471.0          320.0      388.0   
4           0.0     1530.0          628.0            0.0      134.0   

   security_delay  late_aircraft_delay  
0             0.0                425.0  
1             0.0                518.0  
2             0.0                444.0  
3             0.0                218.0  
4             0.0                768.0  

[5 rows x 21 columns]

In [4]:
# Shape of the dataset
print("Shape:", df.shape)

# Column info
df.info()

Shape: (171666, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171666 entries, 0 to 171665
Data columns (total 21 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   year                 171666 non-null  int64  
 1   month                171666 non-null  int64  
 2   carrier              171666 non-null  object 
 3   carrier_name         171666 non-null  object 
 4   airport              171666 non-null  object 
 5   airport_name         171666 non-null  object 
 6   arr_flights          171426 non-null  float64
 7   arr_del15            171223 non-null  float64
 8   carrier_ct           171426 non-null  float64
 9   weather_ct           171426 non-null  float64
 10  nas_ct               171426 non-null  float64
 11  security_ct          171426 non-null  float64
 12  late_aircraft_ct     171426 non-null  float64
 13  arr_cancelled        171426 non-null  float64
 14  arr_diverted         171426 non-null  float64
 1

In [5]:
# Check for missing values
print("\nMissing Values:\n", df.isnull().sum())

# Percent of missing values per column
missing_percent = df.isnull().mean() * 100
missing_percent[missing_percent > 0].sort_values(ascending=False)




Missing Values:
 year                     0
month                    0
carrier                  0
carrier_name             0
airport                  0
airport_name             0
arr_flights            240
arr_del15              443
carrier_ct             240
weather_ct             240
nas_ct                 240
security_ct            240
late_aircraft_ct       240
arr_cancelled          240
arr_diverted           240
arr_delay              240
carrier_delay          240
weather_delay          240
nas_delay              240
security_delay         240
late_aircraft_delay    240
dtype: int64


arr_del15              0.258059
arr_flights            0.139806
carrier_ct             0.139806
weather_ct             0.139806
nas_ct                 0.139806
security_ct            0.139806
late_aircraft_ct       0.139806
arr_cancelled          0.139806
arr_diverted           0.139806
arr_delay              0.139806
carrier_delay          0.139806
weather_delay          0.139806
nas_delay              0.139806
security_delay         0.139806
late_aircraft_delay    0.139806
dtype: float64

In [6]:
# Rows where any column has NaNs
df[df.isnull().any(axis=1)]

year  month carrier                  carrier_name airport  \
3880    2023      6      MQ                     Envoy Air     CLL   
3894    2023      6      MQ                     Envoy Air     EUG   
3898    2023      6      MQ                     Envoy Air     FSD   
3904    2023      6      MQ                     Envoy Air     GRB   
3910    2023      6      MQ                     Envoy Air     HPN   
...      ...    ...     ...                           ...     ...   
168602  2013     10      OO         SkyWest Airlines Inc.     LSE   
169464  2013      9      F9        Frontier Airlines Inc.     MTJ   
170096  2013      9      YV            Mesa Airlines Inc.     FAY   
170150  2013      9      EV      ExpressJet Airlines Inc.     TPA   
171202  2013      8      MQ  American Eagle Airlines Inc.     MSY   

                                             airport_name  arr_flights  \
3880          College Station/Bryan, TX: Easterwood Field          NaN   
3894                       Eugene, OR: Mahlon Sweet Field          NaN   
3898                      Sioux Falls, SD: Joe Foss Field          NaN   
3904    Green Bay, WI: Green Bay Austin Straubel Inter...          NaN   
3910                 White Plains, NY: Westchester County          NaN   
...                                                   ...          ...   
168602                  La Crosse, WI: La Crosse Regional          NaN   
169464              Montrose/Delta, CO: Montrose Regional          NaN   
170096  Fayetteville, NC: Fayetteville Regional/Granni...          NaN   
170150                     Tampa, FL: Tampa International          1.0   
171202  New Orleans, LA: Louis Armstrong New Orleans I...          NaN   

        arr_del15  carrier_ct  weather_ct  ...  security_ct  late_aircraft_ct  \
3880          NaN         NaN         NaN  ...          NaN               NaN   
3894          NaN         NaN         NaN  ...          NaN               NaN   
3898          NaN         NaN         NaN  ...          NaN               NaN   
3904          NaN         NaN         NaN  ...          NaN               NaN   
3910          NaN         NaN         NaN  ...          NaN               NaN   
...           ...         ...         ...  ...          ...               ...   
168602        NaN         NaN         NaN  ...          NaN               NaN   
169464        NaN         NaN         NaN  ...          NaN               NaN   
170096        NaN         NaN         NaN  ...          NaN               NaN   
170150        NaN         0.0         0.0  ...          0.0               0.0   
171202        NaN         NaN         NaN  ...          NaN               NaN   

        arr_cancelled  arr_diverted  arr_delay  carrier_delay  weather_delay  \
3880              NaN           NaN        NaN            NaN            NaN   
3894              NaN           NaN        NaN            NaN            NaN   
3898              NaN           NaN        NaN            NaN            NaN   
3904              NaN           NaN        NaN            NaN            NaN   
3910              NaN           NaN        NaN            NaN            NaN   
...               ...           ...        ...            ...            ...   
168602            NaN           NaN        NaN            NaN            NaN   
169464            NaN           NaN        NaN            NaN            NaN   
170096            NaN           NaN        NaN            NaN            NaN   
170150            0.0           1.0        0.0            0.0            0.0   
171202            NaN           NaN        NaN            NaN            NaN   

        nas_delay  security_delay  late_aircraft_delay  
3880          NaN             NaN                  NaN  
3894          NaN             NaN                  NaN  
3898          NaN             NaN                  NaN  
3904          NaN             NaN                  NaN  
3910          NaN             NaN                  NaN  
...           

Based on the above we can see that most NaN values are coming from airlines not having any flightes to a specific air port during a month. this is important because it gives us a clear direction on how to handle these values. We should be able to drop these rows due to having little impact on the usable data of the data set, infact, the inclusion of such may skew our averages and hide other trends we may be looking for. 

Arguably an easy and simple way to handle the NaN values, however, for this data set it does not make sense to fill in the missing information since the Nan values truely represent data that would only cloud our models judgement. 

In [7]:
#

# Check how many rows we're dropping
null_flight_rows = df[df['arr_flights'].isna()]
print(f"Rows with no arrival flights: {len(null_flight_rows)}")
display(null_flight_rows[['year', 'month', 'carrier', 'airport', 'arr_flights']].head())

# Now drop those rows
df_cleaned = df[~df['arr_flights'].isna()].copy()
print(f"Remaining rows after cleaning: {df_cleaned.shape[0]}")

# Check for missing values
print("\nMissing Values:\n", df_cleaned.isnull().sum())



Rows with no arrival flights: 240


year  month carrier airport  arr_flights
3880  2023      6      MQ     CLL          NaN
3894  2023      6      MQ     EUG          NaN
3898  2023      6      MQ     FSD          NaN
3904  2023      6      MQ     GRB          NaN
3910  2023      6      MQ     HPN          NaN

Remaining rows after cleaning: 171426

Missing Values:
 year                     0
month                    0
carrier                  0
carrier_name             0
airport                  0
airport_name             0
arr_flights              0
arr_del15              203
carrier_ct               0
weather_ct               0
nas_ct                   0
security_ct              0
late_aircraft_ct         0
arr_cancelled            0
arr_diverted             0
arr_delay                0
carrier_delay            0
weather_delay            0
nas_delay                0
security_delay           0
late_aircraft_delay      0
dtype: int64


You can see that there are still NaN values, but what is interesting is that it is found on arrival delays, we cannot simply drop this row, this is actually an important metric due to the fact that there are 203 months where an airline expereince no delays on any of their flights to a airport. 

In [11]:

df_cleaned['arr_del15'] = df_cleaned['arr_del15'].fillna(0)

# Check for missing values
print("\nMissing Values:\n", df_cleaned.isnull().sum())




Missing Values:
 year                   0
month                  0
carrier                0
carrier_name           0
airport                0
airport_name           0
arr_flights            0
arr_del15              0
carrier_ct             0
weather_ct             0
nas_ct                 0
security_ct            0
late_aircraft_ct       0
arr_cancelled          0
arr_diverted           0
arr_delay              0
carrier_delay          0
weather_delay          0
nas_delay              0
security_delay         0
late_aircraft_delay    0
dtype: int64


Now that we have cleaned up and gotten our head around the data, removed the rows that would have provided no significant or useful context and then made sure to fill the arrival delays with 0's to indicate when there were no delays present we should be ready to move onto the next step of the data science method. This was arguably a simple approach but with a nice clean data set to start with, not much pre processing work required. 